local-cluster/tests/local_cluster.rs

```diff
@@ -4480,6 +4480,218 @@ fn test_slot_hash_expiry() {
     );
 }
 
+// This tests vote refreshes on votes that are too old to be in slot hash history. Such votes
+// will never land, so we should instead send a new vote for the tip of the heaviest fork
+// descending from the old vote.
+// It creates validators A, B, and C, (C is only needed for stake distribution), then do this:
+// 1. kill B and C, let A run for a while to create its own fork A
+// 2. kill A, copy ledger of A to B so B has everything until the common ancestor, restart B and
+//    let it run long enough to create a new fork B
+// 3. bring back A and B only, A now has 37% of the stake so its fork is heavier, B has only 35%
+//    so it wants to switch to fork A, but can't do it because of A's stake under switch proof
+//    threshold of 38%. Also, B has voted on fork B, and last_voted < last_landed_vote (we do this
+//    by stopping B after its 4-block leader slot), and last vote is old enough so it's outside
+//    slothash
+// 4. we check that B will be able to vote again and the new vote is further down the B fork.
+#[test]
+#[serial]
+fn test_vote_refresh_outside_slothash() {
+    solana_logger::setup_with_default(RUST_LOG_FILTER);
+    const MAX_SLOT_HASHES: Slot = 8;
+    solana_sdk::slot_hashes::set_entries_for_tests_only(MAX_SLOT_HASHES.try_into().unwrap());
+
+    let slots_per_epoch = 2048;
+    // A has more stake than others, but less than 38% switch proof threshold, so B wants
+    // to switch but could not. C is only added to divide stake better.
+    let node_stakes = vec![
+        37 * DEFAULT_NODE_STAKE,
+        35 * DEFAULT_NODE_STAKE,
+        28 * DEFAULT_NODE_STAKE,
+    ];
+    let node_vote_keys = vec![
+        "3NDQ3ud86RTVg8hTy2dDWnS4P8NfjhZ2gDgQAJbr3heaKaUVS1FW3sTLKA1GmDrY9aySzsa4QxpDkbLv47yHxzr3",
+        "46ZHpHE6PEvXYPu3hf9iQqjBk2ZNDaJ9ejqKWHEjxaQjpAGasKaWKbKHbP3646oZhfgDRzx95DH9PCBKKsoCVngk",
+        "3zsEPEDsjfEay7te9XqNjRTCE7vwuT6u4DHzBJC19yp7GS8BuNRMRjnpVrKCBzb3d44kxc4KPGSHkCmk6tEfswCg",
+    ]
+    .iter()
+    .map(|s| Arc::new(Keypair::from_base58_string(s)))
+    .collect::<Vec<_>>();
+    let num_a_leader_slots = 12;
```

Comment:
Maybe an assert/comment here that `num_a_leader_slots` must be greater than `MAX_SLOT_HASHES` to guarantee that `B`'s vote doesn't land after `B` votes on the last of his 4 blocks